### DISCUSSION TITLE
Senseair S8 sensor not transmitting anything, Micropython on the Pi Pico rp2040

### DISCUSSION THREAD

**Discussion by Axil12** [micropython/micropython#10409] [Hardware & Peripherals] [answered]
Hello there,

I have recently aquired a co2 sensor called Senseair S8. A technical sheet on how to use it can be found [here](http://www.warf.com/download/6337_6449_2.pdf).

I have been trying to read data from the sensor from a Pi Pico 2040. 
My version of micropython is v1.19.1 which can be found [here](https://micropython.org/download/rp2-pico/).
My wiring is as follows : TX pin on Pin 4, RX Pin on Pin 5.
I can see the led from inside the sensor blinking, which tells me that the sensor is correctly powered.

For the code I tried to follow the senseair-s8 library you can find on github [here](https://github.com/ndoornekamp/senseair_s8/blob/main/senseair_s8/senseair_s8.py).
Specifically, the bit of code I used for insipiration is this:

```py
import time
import serial


class SenseairS8:
    
    def __init__(self, port="/dev/ttyS0", baudrate=9600, timeout=.5):
        self.ser = serial.Serial(port, baudrate=baudrate, timeout=timeout)
        self.ser.flushInput()
    
    def co2(self):
        try:
            self.ser.flushInput()
            self.ser.write(b"\xFE\x44\x00\x08\x02\x9F\x25")
            time.sleep(1)
            response = self.ser.read(7)

            high = response[3]
            low = response[4]
            co2 = (high*256) + low
        except Exception as e:
            co2 = None
            print(e)
        
        return co2
```

Now, micropython does not have the pyserial library. From what I understand, I instead need to use machine.UART. So I came up with this bit of code :
```py
from machine import UART, Pin
from time import sleep


def main():
    uart1 = UART(1, baudrate=9600, tx=Pin(4), rx=Pin(5), timeout=500)
    uart1.init(9600, bits=8, parity=None, stop=1)
    uart1.flush()
    sleep(1)
    
    while True:
        uart1.flush()
        sequence_to_send = b'\xFE\x44\x00\x08\x02\x9F\x25'
        uart1.write(sequence_to_send)
        sleep(1)
                
        if uart1.any():
            resp = uart1.read(7)
            high = ord(resp[3])
            low = ord(resp[4])
            co2 = hight * 256 + low
            
            print(co2)
        else:
            print("Nothing to read...")

if __name__ == "__main__":
    main()
```

Now here is my issue : I don't seem to be receiving any data at all. 'uart1.any()' always returns '0', which means there never is anything to read.

I don't quite understand where my mistake is... Any ideas ?

Thank you in advance for your answers !

**Comment by robert-hh** [comment, marked as answer]
Just a few confirmation questions:
- I assume that you connected GND as well. Sometimes people forget it.
- You know that the Pin numbers are GPIO numbers, not board numbers. e.g. GPIO4 is at board pin 6.

Remove the uart.init() statement. It may change values that you have set defore. All parameters you need can be included in the constructor.
You could try to swap RX and TX. Sometines the labelling at devices is reversed, showing the expected connection.

**Reply by robert-hh** [reply]
Another question: reading the linked data sheet, I do not find function code 0x44 which is used in your example. There is a function code 0x04 to read the registers from the device. Did you try the examples from the data sheet. Because is the format is wrong, the device will not respond.

**Reply by Axil12** [reply]
> Just a few confirmation questions:
> 
>     * I assume that you connected GND as well. Sometimes people forget it.
> 
>     * You know that the Pin numbers are GPIO numbers, not board numbers. e.g. GPIO4 is at board pin 6.
> 
> 
> Remove the uart.init() statement. It may change values that you have set defore. All parameters you need can be included in the constructor. You could try to swap RX and TX. Sometines the labelling at devices is reversed, showing the expected connection.

Yes, GND is connected. And the supplied voltage is 5V, so "G+" as it's called on the sensor is connected to VBUS on my Pi Pico.
As for GPIO number and board number, I just look at the printed Pin name to avoid any confusion.
Here is a photo of my wiring :
![20230104_210506](https://user-images.githubusercontent.com/53212285/210640460-46fe7b5f-6add-41c7-b388-fe45bf369e9b.jpg)

Swaping RX and TX pins gave me the following error : 
```py
>>> %Run -c $EDITOR_CONTENT
Traceback (most recent call last):
  File "<stdin>", line 31, in <module>
  File "<stdin>", line 6, in main
ValueError: bad TX pin
>>> 
```

So I think the original Pins I had were fine.

I also got rid of uart1.init(...)

Despite all that, the issue remains the same, there is nothing received from the sensor :(.

Although this time (I suspect that is due to me unplugging and repluggling the Pico), there is a very first read, and it is **b'\x00'**.
After that first successful read, uart1.any() just keeps returning 0 again.
Interestingly, that is also the behavior I get when I just disconnect my TX and RX wires.

**Reply by Axil12** [reply]
> Another question: reading the linked data sheet, I do not find function code 0x44 which is used in your example. There is a function code 0x04 to read the registers from the device. Did you try the examples from the data sheet. Because is the format is wrong, the device will not respond.

This is something that bothered me about the original code snippet I found on github. The sequence of bites to send is strange. But this [Youtube video](https://www.youtube.com/watch?v=395yN1eJKyc) demonstrates that it works somehow.

Anyway, from what I understand from the technical sheet (page 19/23), I should send "_b'\xfe\x04\x00\x03\x00\x01\xd5\xc5_'" for a Co2 reading. Or I should send "_b'\xfe\x04\x00\x00\x00\x04\xe5\xc6'_" for a status + Co2 reading (page 20/23).
But in both cases, I still get no reading from the sensor...

**Reply by robert-hh** [reply]
I did not mean change the code but to swap the wires. The code has to stay the same. I cannot tell from the data sheet which supply voltage is required. You connected the sensor to 5V. What is the voltage level of the TX and RX lines? The Pico uses 3.3V. That may not work with 5V logic.

**Reply by Axil12** [reply]
> I did not mean change the code but to swap the wires. The code has to stay the same. 

Oh my bad.

So I switched the wires, and now it works ! I do get a reading ! Well I get a reading of about 13000 ppm with the code I provided above. So I'm going to assume that's a coding error otherwise I would have dropped to the floor already :D.

**Reply by Axil12** [reply]
>  I get a reading of about 13000 ppm with the code I provided above. So I'm going to assume that's a coding error otherwise I would have dropped to the floor already :D.

I found the bug : 

```py
high = ord(resp[3])
low = ord(resp[4])
```

has to be corrected to :

```py
high = resp[3]
low = resp[4]
```

I then get readings of about 750ppm, which is reasonable for an indoor space.

And with that, I think all of this is resolved !

Thank you very much for the help !

**Reply by cve2022** [reply]
> I did not mean change the code but to swap the wires. The code has to stay the same. I cannot tell from the data sheet which supply voltage is required. You connected the sensor to 5V. What is the voltage level of the TX and RX lines? The Pico uses 3.3V. That may not work with 5V logic.

~~I second @robert-hh 's recommendation, the RP2040 inputs are not 5V tolerant so you should at least add a resistor in series with the RX (1k to limit the current to a few mAmps - but see below) and make sure the TX VOH meets the sensor's minimum VIH.~~

~~The resistor on the RX input is a quick fix and should work for prototyping. Usually IC data sheets specify the maximum input current and it is of a few mAmps or higher, but RP2040's data sheet has no such an information (usually in the absolutely maximum ratings section) so you should use a level converter between the Pico and your sensor for a more permanent solution. This [level converter](https://www.adafruit.com/product/757) (or some similar clone) for example should work for quick prototyping (or even permanent solution) and it must be easy to find.~~

EDIT: out of curiosity I looked at the S8's spec sheet, the UART pins are supplied by 3.3V internally supplied by the sensor, so the concerns above do not apply.
